### PR TITLE
Cache root in non-file-backed buffers

### PR DESCRIPTION
**Pull request by aaronjensen**
-----------------

Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](../CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [x] You've updated the readme (if adding/changing user-visible functionality)

Thanks!

### CODE REVIEW

**Comment by aaronjensen** [comment]
I'm not sure how to add tests for this... let me know if that's a problem. Thanks!

**Comment by bbatsov** [comment]
Looks good to me! Normally the test for this would involve creating some buffer, let-binding a bit of config, running `projectile-project-root` and checking that the cached file name is nil.

File: projectile.el [hunk 1]
```diff
@@ -905,10 +905,10 @@ will return the current directory, otherwise it'd raise an error."
   ;; instead
   (or (cl-subst nil 'none
                 (or (and projectile-cached-buffer-file-name
-                         (equal projectile-cached-buffer-file-name buffer-file-name)
+                         (equal projectile-cached-buffer-file-name (or buffer-file-name 'none))
```
Comment (bbatsov) [inline projectile.el]: Btw, we could have also wrapped this in a `(when buffer-file-name` which might have been more clear, but this came to my mind only after I merged this. Anyways, not a big deal.

Comment (aaronjensen) [reply projectile.el]: Wrapped which? I tried wrapping the whole thing in that previously and it broke some functionality for me.

**Comment by aaronjensen** [comment]
Also, thanks for the merge!

**Pull request merged**